cmd/object-handlers.go

```diff
@@ -1064,6 +1064,8 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 	bucket := vars["bucket"]
 	object := vars["object"]
 
+	r.Body = &detectDisconnect{r.Body, r.Context().Done()}
```

Comment:
Can we not generalize this for as a generic handler for all `r.Body = &detectDisconnect{r.Body, r.Context().Done()}`

Reply:
This type of change is not useful for other calls as most of them don't have a body. Even if some do, the body gets read early in signature verification code for sha256.

PutObjectPart is an exception, in which case client timingout will not cause any difference as client is going to send completemultipart() in the end to finalize things.

Better restrict this change to just PutObject. (I feel it's a little hacky to replace http.Request's feilds, even though we do it here and sha256 verification for convenience)

Reply:
The reason is we need a generic solution for locking timeouts. It is not just Put object behavior we should fix situation of being able to not proceed with the handler when the client times out. If it's based on r.Body it only addresses Put use case, you need to handle POST upload scenario as well.

Even in case of other operations we should have the same strategy where an explicit client timeout should invariably cancel the locks and returns the handler. 

For now this fix is not complete..

Reply:
I would infact prefer we expand lsync to take context and then wait on the ctx.Done() such that locking can be returned. It would be a more idiomatic fix.

Reply:
> I would infact prefer we expand lsync to take context and then wait on the ctx.Done() such that locking can be returned. It would be a more idiomatic fix.

No this will not solve the problem either. For ex. if the client times out after we get the lock and before we respond with 200OK, we can't undo the operation.

Ideally clients should wait for responses before it can definitively decide if the operation succeeded or not. Otherwise the behavior is undefined as explained in the above situation.

But it is a desirable behavior for the server to make best efforts to abort the operation when it knows that the client has timed out.

The change in this PR is needed even if we implement context-cancelling the locks. Because the client can timeout after we get the lock and when reads are happening on `r.Body` from the server side.

We can implement aborting the locks in a different PR as it is more involved.

Reply:
It really doesn't matter after lock is acquired, if the client timeouts before it sends all the body object will not be created.

Reply:
Content-Length is there for this reason alone. This PR doesn't solve the required issue IMO, we should fix the locking behavior.

 r.Body is already proof of incorrect objects being created lesser or greater than required content length.

Reply:
> Content-Length is there for this reason alone.

`Content-Length` will address the problem if the client has not sent all the data. But if the client has sent all the data and then times out (after acquiring the lock) then the fix in this PR is needed to detect the client disconnection.

> This PR doesn't solve the required issue IMO, we should fix the locking behavior.

The PR was tested with the customer's test case and it fixes the issue. i.e after lock is acquired, it tries to read the data and fails.

This PR in particular addresses this specific situation:
1) client has sent full data and the data is available in r.Body (bufio reader)
2) lock acquired
3) client times out and disconnects

In this situation `Content-Length` will not protect, because r.Body's bufio-reader would contain all the data sent by the client.

Probably I can explain it in person.

Locking cancellation can be fixed in another PR (That PR will be to address the situation where client timesout when we are blocked on lock() so it can fail quicker).

Reply:
Yes but we don't need to fix this, because if you fix locking behavior everything fixes automatically. The fundamental issue is locks were granted for disconnected clients.

In my understanding this PR is fixing a niche situation where Content-Length is entirely buffered. But in reality what happened was locks were granted for timedout clients.

Another reason is relinquishing locking calls when client timesout is important from a resource consumption point of view on server side. Overall a better and cleaner behavior.